Add support for DuckDuckGo

Veritania opened the pull request:
Allows for search queries done via [DuckDuckGo](https://duckduckgo.com/) to replace the `poe` keyword with `site:poewiki.net`, extending the existing google functionality to another search engine.

Tested successfully on Mozilla Firefox 91.0.2

pm5k commented:
Hello,

Please can you merge 7ebd71e into your branch since we made some changes to the query parsing. @tonytangio has corrected the parsing to ensure we don't "hijack" non-wiki `poe` queries. Your PR should reflect this, and then, I think I am happy to merge it once corrected. :)

tonytangio commented on src/background/redirects.js:
```diff
@@ -29,4 +29,16 @@ function redirectFromGoogle(requestDetails) {
+
+    return {
+        redirectUrl: `https://www.duckduckgo.com/?q=site:poewiki.net+${searchQuery}`,
+    }
```
(After merging/rebasing from master)

Note how the implementation of both this and `redirectFromGoogle` are identical - you should move these to a common `searchQueryFromRequest` function and remove the duplication entirely.

This will make future search engine support trivial to add.

tonytangio replied:
Also, please wait to merge/rebase from master, after #10 has been merged, to minimise any other conflicts. Thanks :)

pm5k commented:
@Veritania  just getting in touch with you regarding this PR and the progress. If you believe you may not have the time to implement this fully based on the comments above, can you let us know and see if you're comfortable with this being taken over from you on our end so we can pack this change into the future 1.0.7 version? Thanks.

Veritania commented:
@pm5k Hi, I'll try to give it a go this weekend.

Veritania commented:
Hi @pm5k 
I've merged in your latest master and updated my branch to be more generic-use and less duplication now, as requested by @tonytangio 

I've also attempted to update the tests, to be easier to add in a new search engine and run the same queries against it. 
I'm open to feedback, and thank you for your time :+1: 

These changes should make it less work to include more search engines in the future.

tonytangio commented on src/background/redirects.spec.js:
```diff
@@ -20,61 +20,85 @@ describe("Fandom redirect", () => {
+    ]
+
+    //The base redirect result URL to expect from each search engine
+    const redirectHost = new Map()
```
1. Why not just use a JS object? Map isn't necessary here as you're not dynamically changing these keys+values

2. nit - containers like these maps and objects should be plural (because they hold multiple elements)

3. nit - the keys of this container represent the hosts, but the *values* of this map are URLs that's more than just the host (not sure what I'd call them, perhaps "base"s?)

Here's how I'd approach this (you're welcome to disagree!):

```js
const redirectBaseUrls = {
  google: "https://www.google.com/search?q=site:poewiki.net+",
  duckduckgo: "https://www.duckduckgo.com/?q=site:poewiki.net+"
}
// .get becomes
redirectBaseUrls[searchEngine]
```

tonytangio reviewed (commented):
👍 Nice work, just have a few small comments and I'm happy to approve after.

tonytangio commented on src/background/redirects.spec.js:
```diff
@@ -20,61 +20,85 @@ describe("Fandom redirect", () => {
+            it.each([
+                {
+                    url: urlHost + "q=poe+faster+attacks",
+                    expected: redirectHost.get(searchEngine) + "faster+attacks"
```
nit - minor duplication removal/optimization:
```suggestion
                    expected: redirectBaseUrl + "faster+attacks"
```

Under line 44
```js
const redirectBaseUrl = redirectBaseUrls[searchEngine]
```

Veritania commented:
@tonytangio  Thank you for the feedback. I've implemented the changes you've suggested.
I also added a default to the switch now in 50838c5 . I like that better than having the chance of using a null value in the switch. 

Even if we should never hit the default, you might not agree with it redirecting to ddg instead of google. Privacy vs familiarity in case a user hits the default and suddenly see ddg, a website that might be unknown to them.

tonytangio commented:
> Even if we should never hit the default, you might not agree with it redirecting to ddg instead of google.

Yeah, I'm leaning towards Google since it's easily the majority choice. Should we ever hit the default case, we don't want most people getting surprised with a different search engine (one they might not even have heard of) and losing trust in the extension.

tonytangio commented on src/background/redirects.js:
```diff
@@ -25,10 +24,26 @@ function redirectFromGoogle(requestDetails) {
+        case "www.duckduckgo.com":
+            searchEngine = 'https://www.duckduckgo.com/?q=site:poewiki.net+'
+            break;
+        default: searchEngine = 'https://www.duckduckgo.com/?q=site:poewiki.net+'
```
nit: you could just initialise `searchEngine` with the default rather than setting it here, you could even rearrange the fall-throughs like this:

```js
        default:
        case "duckduckgo.com":
        case "www.duckduckgo.com":
            searchEngine = 'https://www.duckduckgo.com/?q=site:poewiki.net+'
 ```

Veritania replied:
Fixed in d512a69 , our fallback is google now - I've removed the default all-together as it serves no purpose anymore.
